amethyst_error/src/lib.rs

```diff
@@ -0,0 +1,422 @@
+//! Contains the `Error` type and company as used by Amethyst.
+//!
+//! **Note:** This type is not intended to be used outside of Amethyst.
+//! If you are integrating a crate from amethyst to use this, it is recommended that you treat this
+//! type as an opaque [`std::error::Error`].
+//!
+//! [`std::error::Error`]: https://doc.rust-lang.org/std/error/trait.Error.html
+
+// Parts copied from failure:
+// https://github.com/rust-lang-nursery/failure
+
+#![warn(missing_docs, rust_2018_idioms, rust_2018_compatibility)]
+
+#[cfg(all(feature = "backtrace", feature = "std"))]
+extern crate backtrace;
+
+use self::internal::Backtrace;
+use std::borrow::Cow;
+use std::error;
+use std::fmt;
+use std::result;
+
+/// Internal parts of `Error`.
+#[derive(Debug)]
+struct Inner<T: ?Sized> {
+    source: Option<Box<Error>>,
+    #[cfg(all(feature = "backtrace", feature = "std"))]
+    backtrace: Option<Backtrace>,
+    error: T,
+}
+
+/// The error type used by Amethyst.
+///
+/// Wraps error diagnostics like messages and other errors, and keeps track of causal chains and
+/// backtraces.
+pub struct Error {
+    inner: Box<Inner<dyn error::Error + Send + Sync>>,
+}
+
+impl Error {
+    /// Default constructor for our error types.
+    ///
+    /// Wraps anything that is an error in a box.
+    pub fn new<E>(error: E) -> Self
+    where
+        E: 'static + error::Error + Send + Sync,
+    {
+        Self {
+            inner: Box::new(Inner {
+                source: None,
+                #[cfg(all(feature = "backtrace", feature = "std"))]
+                backtrace: self::internal::bt::new(),
+                error: Box::new(error),
+            }),
+        }
+    }
+
+    /// Construct a new error from a string.
+    pub fn from_string<M>(message: M) -> Self
+    where
+        M: Into<Cow<'static, str>>,
+    {
+        /// Wrapper for string errors.
+        #[derive(Debug)]
+        struct StringError(Cow<'static, str>);
+
+        impl fmt::Display for StringError {
+            fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+                self.0.fmt(fmt)
+            }
+        }
+
+        impl error::Error for StringError {}
+
+        Self {
+            inner: Box::new(Inner {
+                source: None,
+                #[cfg(all(feature = "backtrace", feature = "std"))]
+                backtrace: self::internal::bt::new(),
+                error: Box::new(StringError(message.into())),
+            }),
+        }
+    }
+
+    /// Get backtrace.
+    pub fn backtrace(&self) -> Option<&Backtrace> {
+        #[cfg(all(feature = "backtrace", feature = "std"))]
+        {
+            self.backtrace.as_ref()
+        }
+
+        #[cfg(not(all(feature = "backtrace", feature = "std")))]
+        {
+            None
+        }
+    }
+
+    /// Get the source of the error.
+    ///
+    /// # Examples
+    ///
+    /// The only way to set the source is through [`ResultExt`](ResultExt) using
```

Comment:
We should provide a way to create `Error` with source without `ResultExt`.
Simple `with_context` method that works same way as `ResultExt` would do.

Reply:
:+1:

Reply:
Added `Error::with_source`.